velox/functions/sparksql/aggregates/tests/CorrelationAggregationTest.cpp

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "velox/common/base/tests/GTestUtils.h"
+#include "velox/exec/tests/utils/PlanBuilder.h"
+#include "velox/functions/lib/aggregates/tests/AggregationTestBase.h"
+#include "velox/functions/sparksql/aggregates/Register.h"
+
+using namespace facebook::velox::exec::test;
+using namespace facebook::velox::functions::aggregate::test;
+
+namespace facebook::velox::functions::aggregate::sparksql::test {
+
+namespace {
+
+class CorrelationAggregationTest : public AggregationTestBase {
+ protected:
+  void SetUp() override {
+    AggregationTestBase::SetUp();
+    allowInputShuffle();
+    registerAggregateFunctions("spark_");
+  }
+};
+
+TEST_F(CorrelationAggregationTest, precision) {
+  // Construct data to make covar = 0.5, both stddev = 0.5.
+  auto data = makeRowVector({
+      makeFlatVector<double>({0, 1}),
+      makeFlatVector<double>({0, 1}),
+  });
+  auto expected =
+      makeRowVector({makeFlatVector<int64_t>(std::vector<int64_t>({1}))});
+
+  // Add an extra cast to test the result is not 0.999...
+  testAggregations(
+      {data},
+      {},
+      {"spark_corr(c0, c1)"},
+      {"cast(a0 as BIGINT)"},
```

Comment:
This cast computation itself is not stable, both 0 or 1 are acceptable results.  M/(sqrt(A)sqrt(B)) is numerically safer than M/sqrt(AB) and should be preferred.

Reply:
In sparksql, cast from double to int always truncate the fractional part, so `cast(0.99 as int)` gets 0, `cast(1.0 as int)` gets 1, which are different results. We should keep consistent with spark.

Reply:
Truncating on floating point number to convert it to integer enlarges the error by $1/\epsilon$ so is not stable.  The calculation here is deemed to be inaccurate and have nothing to do with how `corr` is calculated

Reply:
I agree it's unstable but it's the current implementation in Spark. We need to align with Spark's implementation and get the same result. Otherwise customer will see mismatched result in Velox and Spark. We can remove the logic once Spark use Presto's implementation. 

The PR is to required to fix a Spark UT.

Reply:
@FelixYBW Given that Spark's implementation is unstable, I expect customers already getting inconsistent results.  Have you opened an issue with Spark to fix this implementation? It doesn't seem to make sense to implement Spark's bugs in Velox.

Reply:
No customer report the inconsistent results. The fix is to pass a Spark UT. @marin-ma Is the UT failed due to inconsistent result?

Reply:
@FelixYBW Yes, one of group-by Spark UT fails. And #4917 was reported by the customer.

Reply:
In the customer reported issue Velox get more accurate result. So it's more spark's bug.

@marin-ma Let's close the PR, change the expected value as Velox's in Spark UT, then mark the difference in Gluten's doc. Later let's submit a PR to Spark community.

Thank you @mbasmanova  and @Yuhta